### ISSUE TITLE
Implement Recurrent Layers

### ISSUE THREAD

**Issue by adam-r-kowalski** [tensorflow/swift-apis#52]
It would be really nice to have a set of recurrent layers including LSTMs and GRUs as part of the core layers API. Ideally we can parameterize the activation functions for the various gates rather than hard coding in the tanh and sigmoid functions.

**Comment by rxwei** [comment]
Definitely, they are on our roadmap. Thanks for opening an issue about this.

**Comment by rxwei** [comment]
FWIW, here's an example RNN implemented using the Swift of TensorFlow Deep Learning Library: https://gist.github.com/rxwei/ce6644efad8f229651050e096c05ccbb.

Most people are often stuck at these two things on their first try:
* Defining a layer with multiple inputs and outputs.
    * Answer: Define a `struct Input : Differentiable { ... }` and `struct Output : Differentiable { ... }` in your layer with however many arity you need. As to variable-sized inputs/outputs, [`Array` will conform to `Differentiable`](https://github.com/apple/swift/pull/23183) soon.
* Backpropagation through time without differentiating control flow.
    * Answer: Control flow differentiation is not required for BPTT. Each `Layer` has a `Backpropagator` member type. Use `appliedForBackpropagation` to evaluate the forward pass and get a back propagator, and accumulate backpropagators in an array.

You can find how these are done in my gist. If anyone's interested in taking a stab and implement RNN cells, let me know!

**Comment by rxwei** [comment]
Update: #80 added the `RNNCell` protocol and #71 added `SimpleRNN` and `LSTM`. We still need to add GRU and hopefully NTM.

**Comment by dhasl002** [comment]
I'll take a shot at GRU and NTM

**Comment by dhasl002** [comment]
Update: #417 has created to implement GRU cells. I intend to move on to NTM next.
